### ISSUE TITLE
Cant execute tremc

### ISSUE THREAD

**Issue by vredesbyyrd** [tremc/tremc#15]
I recently had to reinstall my root partition because btrfs file corruption. Prior to that i had no issues running tremc. When i try to run it now i get the error(s) 

raceback (most recent call last):
  File "/usr/bin/tremc", line 3702, in <module>
    Interface()
  File "/usr/bin/tremc", line 985, in __init__
    self.run()
  File "/usr/bin/tremc", line 1079, in run
    self.draw_stats()
  File "/usr/bin/tremc", line 2615, in draw_stats
    self.draw_global_rates()
  File "/usr/bin/tremc", line 2694, in draw_global_rates
    self.screen.addch(curses.ACS_UARROW, curses.A_REVERSE)
_curses.error: addch() returned ERR

I am no python wiz, so this could be a configuration issue with my environment, but i can execute other python3 scripts no issue. I also have no trouble running transmission-remote-cli. So i think that rules out permission issues. Any pointers would be greatly appreciated.

**Comment by kozross** [comment]
I'm having this exact problem with an install of tremc.

**Comment by vredesbyyrd** [comment]
I discovered something strange on accident. I have been using transmission-remote-cli in place of tremc, since i cant get tremc to work currently. 

If a torrent is already actively downloading with transmission-daemon, you can enter tremc. It doesnt give me any errors. Only if there is an active torrent though. Who knows....

Im using gnome 3.26-2 with arch/kernel-4.14.9-1 & xorg-server 1.16-2.

**Comment by louipc** [comment]
On Mon 01 Jan 2018 12:38 -0800, vredesbyyrd wrote:
> I discovered something strange on accident. I have been using
> transmission-remote-cli in place of tremc, since i cant get tremc to work
> currently.
>
> If a torrent is already actively downloading with transmission-daemon, you can
> enter tremc. It doesnt give me any errors. Only if there is an active torrent
> though. Who knows....
>
> Im using gnome 3.26-2 with arch/kernel-4.14.9-1 & xorg-server 1.16-2.

Operating system is relevant, transmission-cli, transmission-daemon and
python versions could be relevant for debugging.

Thats really strange.
So if you pause all torrents, exit tremc and relaunch it crashes again?

**Comment by vredesbyyrd** [comment]
>Thats really strange.
So if you pause all torrents, exit tremc and relaunch it crashes again? 

I just tested tremc some more. 

So as long as a torrent is added, tremc will start. Whether Its active, idle, or paused is irrelevant. If i remove the torrent, exit tremc, try to enter it will crash.  

> Operating system is relevant, transmission-cli, transmission-daemon and
python versions could be relevant for debugging.

~Arch Linux 
~transmission-cli  version-2.92-7 
~transmission-remote-cli-git  version-1.7.1.3
~python2  version-2.7.14-1 
~python3  version-3.6.4-1 

I start the daemon manually with 'transmission-daemon'. I do not use the systemd daemon service. I tried with the service but it didnt change anything regarding tremc not starting. The daemon is part of the transmission-cli package.

**Comment by louipc** [comment]
On Tue 02 Jan 2018 01:24 +0000, vredesbyyrd wrote:
> >Thats really strange.
> So if you pause all torrents, exit tremc and relaunch it crashes again?
>
> I just tested tremc some more.
>
> So as long as a torrent is added, tremc will start. Whether Its active, idle,
> or paused is irrelevant. If i remove the torrent, exit tremc, try to enter it
> will crash.

Thanks I can reproduce the crash. I should be able to debug.

> ~Arch Linux
> ~transmission-cli  version-2.92-7
> ~python3  version-3.6.4-1

I have the same.

**Comment by vredesbyyrd** [comment]
Nice. Sounds good, thanks.

**Comment by gangelop** [comment]
I did a git bisect and it turns out that commit e06d08d8d76aa0559593ffc1188f4a90100cdbdb introduced this bug.

y tho?

I'll have to look more into it later.

**Comment by gangelop** [comment]
I've found out which return value causes the problem.
`scale_bytes(0)` used to return the string `"0K"`.
`scale_bytes(0)` now returns the string `"0.0K"` and this breaks something in the ncurses code which produces the traceback reported originally.

Overall, I don't see anything *too unreasonable* with the "scale_bytes". I'd like to find out why the ncurses code is so sensitive to this change, which seems more unreasonable to me.

**Comment by gangelop** [comment]
OK, the problem starts earlier.
The reason we get `curses.error: addch() returned ERR` at `self.screen.addch(curses.ACS_UARROW, curses.A_REVERSE)` is *likely* because it's trying to draw that ACS_UARROW character out of the window.

The reason it it presumably ran out of space is because in this bid of code:

```
self.screen.addstr(scale_bytes(self.stats['downloadSpeed']).rjust(self.rateDownload_width),
                   curses.color_pair(self.colors.id('download_rate'))
                   + curses.A_REVERSE + curses.A_BOLD)
```

the `rateDownload_width` is too small. It's expecting the width of 0K but it gets 0.0K instead.

This is all very silly so I would like to review it again to understand how the width is handled in all cases in general.
For now I've pushed something that seems like a reasonable fix to me in branch issue15, commit 058f437.

**Comment by vredesbyyrd** [comment]
@gangelop 

Thanks for debugging this. Its beyond me what you did to fix it, but I just tried the new commit and it seems to have resolved the problems I was seeing. Smooth sailing.

**Comment by gangelop** [comment]
cool! Thanks for testing. To rephrase what I said in the commit: I'm quite sure that this fixes the problem. I'm just not sure if it might introduce some other bug. I want to understand how those widths are used overall to make sure we avoid introducing another bug.

**Comment by vredesbyyrd** [comment]
No prob, appreciate you getting this back in working order. Ill post If I see any strange behavior / bugs. Happy to be using tremc again rather than the python2 version.

**Comment by louipc** [comment]
On Mon 02 Jul 2018 09:01 -0700, George Angelopoulos wrote:
> The reason we get `curses.error: addch() returned ERR` at
> `self.screen.addch(curses.ACS_UARROW, curses.A_REVERSE)` is *likely* because
> it's trying to draw that ACS_UARROW character out of the window.
> 
> the `rateDownload_width` is too small. It's expecting the width of 0K but it
> gets 0.0K instead.
> 
> This is all very silly so I would like to review it again to understand how
> the width is handled in all cases in general.
> For now I've pushed something that seems like a reasonable fix to me in branch
> issue15, commit 058f437.

Thanks for your work. I wanted to have simplified code and consistent interface
(field width), just hacking it together though.

**Comment by gangelop** [comment]
Ok. I've written a fix I'm happy with in 0cb919b
Will revisit this tomorrow and if I don't notice anything odd I'll merge it.
